src/faction_camp.cpp

```diff
@@ -1962,132 +2066,85 @@ bool basecamp::survey_return( npc &p )
+bool talk_function::camp_farm_return( npc &p, const std::string &task, const tripoint &omt_tgt,
+                                      farm_ops op )
 {
-    std::string dir = camp_direction( task );
-    const tripoint omt_pos = p.global_omt_location();
-    tripoint omt_trg;
-    std::vector<std::pair<std::string, tripoint>> om_expansions = om_building_region( p, 1, true );
-    for( const auto &e : om_expansions ) {
-        if( dir == om_simple_dir( omt_pos, e.second ) ) {
-            omt_trg = e.second;
-            break;
-        }
+    npc_ptr comp = companion_choose_return( p, task, calendar::turn );
+    if( comp == nullptr ) {
+        return false;
     }
-    int harvestable = 0;
-    int plots_empty = 0;
-    int plots_plow = 0;
+
+    /* copy pasted from camp_farm_act because passing pointers does't work */
     bool plow = op & farm_ops::plow;
     bool plant = op & farm_ops::plant;
     bool harvest = op & farm_ops::harvest;
 
-    //bay_json is what the are should look like according to jsons
-    tinymap bay_json;
-    bay_json.generate( omt_trg.x * 2, omt_trg.y * 2, omt_trg.z, calendar::turn );
-    //bay is what the area actually looks like
-    tinymap bay;
-    bay.load( omt_trg.x * 2, omt_trg.y * 2, omt_trg.z, false );
-    for( int x = 0; x < 23; x++ ) {
-        for( int y = 0; y < 23; y++ ) {
-            //Needs to be plowed to match json
-            if( bay_json.ter( x, y ) == ter_str_id( "t_dirtmound" ) &&
-                ( bay.ter( x, y ) == ter_str_id( "t_dirt" ) ||
-                  bay.ter( x, y ) == ter_str_id( "t_grass" ) )
-                && bay.furn( x, y ) == furn_str_id( "f_null" ) ) {
-                plots_plow++;
-            }
-            if( bay.ter( x, y ) == ter_str_id( "t_dirtmound" ) &&
-                bay.furn( x, y ) == furn_str_id( "f_null" ) ) {
-                plots_empty++;
-            }
-            if( bay.furn( x, y ) == furn_str_id( "f_plant_harvest" ) &&
-                !bay.i_at( x, y ).empty() ) {
-                const item &seed = bay.i_at( x, y )[0];
-                if( seed.is_seed() ) {
-                    harvestable++;
-                }
-            }
-        }
-    }
-    time_duration work = 0_minutes;
-    if( harvest ) {
-        work += 3_minutes * harvestable;
-    }
-    if( plant ) {
-        work += 1_minutes * plots_empty;
-    }
-    if( plow ) {
-        work += 5_minutes * plots_plow;
-    }
-
-    npc_ptr comp = companion_choose_return( p, task, calendar::turn - work );
-    if( comp == nullptr ) {
-        return false;
-    }
-
-    std::vector<item *> seed_inv = comp->companion_mission_inv.items_with( []( const item & itm ) {
+    const auto is_dirtmound = []( const tripoint & pos, tinymap & bay1, tinymap & bay2 ) {
+        const ter_str_id &dirtmound = ter_str_id( "t_dirtmound" );
+        const furn_str_id &f_null = furn_str_id( "f_null" );
+        return ( bay1.ter( pos ) == dirtmound ) && ( bay2.furn( pos ) == f_null );
```

Comment:
```suggestion
        return bay1.ter( pos ) == ter_str_id( "dirtmound" ) && !bay2.has_furn( pos );
```

(Why are you are creating those `string_id` variables when you use them only once anyway?)

Reply:
Copying bad code from acidia.